executor/importer/chunk_process.go

```diff
@@ -0,0 +1,300 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package importer
+
+import (
+	"context"
+	"io"
+	"time"
+
+	"github.com/docker/go-units"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/br/pkg/lightning/backend"
+	"github.com/pingcap/tidb/br/pkg/lightning/backend/kv"
+	"github.com/pingcap/tidb/br/pkg/lightning/checkpoints"
+	"github.com/pingcap/tidb/br/pkg/lightning/common"
+	"github.com/pingcap/tidb/br/pkg/lightning/log"
+	"github.com/pingcap/tidb/br/pkg/lightning/mydump"
+	verify "github.com/pingcap/tidb/br/pkg/lightning/verification"
+	"github.com/pingcap/tidb/tablecodec"
+	"github.com/tikv/client-go/v2/tikv"
+	"go.uber.org/zap"
+)
+
+// constants, make it a variable for test
+var (
+	maxKVQueueSize         = 32             // Cache at most this number of rows before blocking the encode loop
+	MinDeliverBytes uint64 = 96 * units.KiB // 96 KB (data + index). batch at least this amount of bytes to reduce number of messages
+	// see default for tikv-importer.max-kv-pairs
+	MinDeliverRowCnt = 4096
+)
+
+type deliveredRow struct {
+	kvs    *kv.Pairs // if kvs is nil, this indicated we've got the last message.
+	offset int64
+	rowID  int64
+}
+
+type deliverResult struct {
+	totalDur time.Duration
+	err      error
+}
+
+type deliverKVBatch struct {
+	dataKVs  kv.Pairs
+	indexKVs kv.Pairs
+
+	dataChecksum  *verify.KVChecksum
+	indexChecksum *verify.KVChecksum
+
+	codec tikv.Codec
+}
+
+func newDeliverKVBatch(codec tikv.Codec) *deliverKVBatch {
+	return &deliverKVBatch{
+		dataChecksum:  verify.NewKVChecksumWithKeyspace(codec),
+		indexChecksum: verify.NewKVChecksumWithKeyspace(codec),
+		codec:         codec,
+	}
+}
+
+func (b *deliverKVBatch) reset() {
+	b.dataKVs.Clear()
+	b.indexKVs.Clear()
+	b.dataChecksum = verify.NewKVChecksumWithKeyspace(b.codec)
+	b.indexChecksum = verify.NewKVChecksumWithKeyspace(b.codec)
+}
+
+func (b *deliverKVBatch) size() uint64 {
+	return b.dataChecksum.SumSize() + b.indexChecksum.SumSize()
+}
+
+func (b *deliverKVBatch) add(kvs *kv.Pairs) {
+	for _, pair := range kvs.Pairs {
+		if pair.Key[tablecodec.TableSplitKeyLen+1] == 'r' {
```

Comment:
Maybe we can use `tablecodec.IsRecordKey`

Reply:
[fix comments](https://github.com/pingcap/tidb/pull/42817/commits/c0dbaee38127111703e5fdd5ef1b3693ee1cf6be)